### PR TITLE
Update the unique index for the systemtag table

### PR DESCRIPTION
**Pull request by sharidas**
Update the unique index for the systemtag table.

Signed-off-by: Sujith H <sharidasan@owncloud.com>

## Description

Update the unique index for the systemtag table.

## Related Issue

- Fixes <issue_link>

## Motivation and Context

If the unique index is not updated with `assignable`, then inserts to the table were causing problems.

## How Has This Been Tested?

- Download owncloud http://download.owncloud.org/community/owncloud-8.2.11.tar.bz2 and install it.
- Download owncloud http://download.owncloud.org/community/owncloud-daily-stable10-qa.tar.bz2 for the upgrade
- When upgraded from 8.2.11 to 10 below are the upgrade log captured from the console:
```
➜  owncloud4 ./occ upgrade                                                 
Cannot load Xdebug - it was already loaded
ownCloud or one of the apps require upgrade - only a limited number of commands are available
You may use your browser or the occ upgrade command to do the upgrade
2018-11-09T11:00:26+00:00 Set log level to debug
2018-11-09T11:00:26+00:00 Repair step: Repair MySQL database engine
2018-11-09T11:00:26+00:00 Repair info: Not a mysql database -> nothing to do
2018-11-09T11:00:26+00:00 Repair step: Repair MySQL collation
2018-11-09T11:00:26+00:00 Repair info: Not a mysql database -> nothing to do
2018-11-09T11:00:26+00:00 Repair step: Repair SQLite autoincrement
2018-11-09T11:00:26+00:00 Repair step: Repair orphaned reshare
2018-11-09T11:00:26+00:00 Repair step: Repair duplicate entries in oc_lucene_status
2018-11-09T11:00:26+00:00 Repair info: lucene_status table does not exist -> nothing to do
2018-11-09T11:00:26+00:00 Repair step: Upgrade app code from the marketplace
2018-11-09T11:00:26+00:00 Repair info: Enabling market app to assist with update
2018-11-09T11:00:26+00:00 Repair info: Using market to update existing apps
2018-11-09T11:00:26+00:00 Repair info: Attempting to update the following missing apps from market: activity, files_pdfviewer, files_texteditor, gallery, templateeditor
2018-11-09T11:00:26+00:00 Repair info: Fetching app from market: activity
2018-11-09T11:00:45+00:00 Repair info: Fetching app from market: files_pdfviewer
2018-11-09T11:01:10+00:00 Repair info: Fetching app from market: files_texteditor
2018-11-09T11:01:28+00:00 Repair info: Fetching app from market: gallery
2018-11-09T11:01:40+00:00 Repair info: Fetching app from market: templateeditor
2018-11-09T11:01:51+00:00 Repair info: Attempting to update the following existing compatible apps from market: activity, files_pdfviewer, files_texteditor, gallery, templateeditor
2018-11-09T11:01:51+00:00 Repair info: Fetching app from market: dav
2018-11-09T11:01:51+00:00 Repair info: App (dav) is not installed
2018-11-09T11:01:51+00:00 Repair info: Fetching app from market: federatedfilesharing
2018-11-09T11:01:51+00:00 Repair info: App (federatedfilesharing) is not installed
2018-11-09T11:01:51+00:00 Repair info: Fetching app from market: files
2018-11-09T11:01:54+00:00 Repair info: App (files) is not known at the marketplace.
2018-11-09T11:01:54+00:00 Repair info: Fetching app from market: files_external
2018-11-09T11:01:54+00:00 Repair info: App (files_external) is not installed
2018-11-09T11:01:54+00:00 Repair info: Fetching app from market: files_sharing
2018-11-09T11:01:57+00:00 Repair info: App (files_sharing) is not known at the marketplace.
2018-11-09T11:01:57+00:00 Repair info: Fetching app from market: files_trashbin
2018-11-09T11:02:01+00:00 Repair info: App (files_trashbin) is not known at the marketplace.
2018-11-09T11:02:01+00:00 Repair info: Fetching app from market: files_versions
2018-11-09T11:02:04+00:00 Repair info: App (files_versions) is not known at the marketplace.
2018-11-09T11:02:04+00:00 Repair info: Fetching app from market: firstrunwizard
2018-11-09T11:02:07+00:00 Repair info: App (firstrunwizard) is not known at the marketplace.
2018-11-09T11:02:07+00:00 Repair info: Fetching app from market: notifications
2018-11-09T11:02:10+00:00 Repair info: App (notifications) is not known at the marketplace.
2018-11-09T11:02:10+00:00 Repair info: Fetching app from market: provisioning_api
2018-11-09T11:02:13+00:00 Repair info: App (provisioning_api) is not known at the marketplace.
2018-11-09T11:02:13+00:00 Repair info: App was not updated: dav
2018-11-09T11:02:13+00:00 Repair info: App was not updated: federatedfilesharing
2018-11-09T11:02:13+00:00 Repair info: App was not updated: files
2018-11-09T11:02:13+00:00 Repair info: App was not updated: files_external
2018-11-09T11:02:13+00:00 Repair info: App was not updated: files_sharing
2018-11-09T11:02:13+00:00 Repair info: App was not updated: files_trashbin
2018-11-09T11:02:13+00:00 Repair info: App was not updated: files_versions
2018-11-09T11:02:13+00:00 Repair info: App was not updated: firstrunwizard
2018-11-09T11:02:13+00:00 Repair info: App was not updated: notifications
2018-11-09T11:02:13+00:00 Repair info: App was not updated: provisioning_api
2018-11-09T11:02:13+00:00 Updating database schema
2018-11-09T11:02:15+00:00 Updated database
2018-11-09T11:02:15+00:00 Updating <files_pdfviewer> ...
2018-11-09T11:02:15+00:00 Updated <files_pdfviewer> to 0.9.0
2018-11-09T11:02:15+00:00 Updating <files_texteditor> ...
2018-11-09T11:02:15+00:00 Updated <files_texteditor> to 2.2.1
2018-11-09T11:02:15+00:00 Updating <gallery> ...
2018-11-09T11:02:15+00:00 Updated <gallery> to 16.1.0
2018-11-09T11:02:15+00:00 Updating <templateeditor> ...
2018-11-09T11:02:15+00:00 Updated <templateeditor> to 0.3.1
2018-11-09T11:02:15+00:00 Updating <files> ...
2018-11-09T11:02:15+00:00 Updated <files> to 1.5.1
2018-11-09T11:02:15+00:00 Updating <activity> ...
2018-11-09T11:02:15+00:00 Updated <activity> to 2.3.8
2018-11-09T11:02:15+00:00 Updating <files_sharing> ...
2018-11-09T11:02:16+00:00 Updated <files_sharing> to 0.11.0
2018-11-09T11:02:16+00:00 Updating <files_trashbin> ...
2018-11-09T11:02:16+00:00 Updated <files_trashbin> to 0.9.1
2018-11-09T11:02:16+00:00 Updating <files_versions> ...
2018-11-09T11:02:16+00:00 Updated <files_versions> to 1.3.0
2018-11-09T11:02:16+00:00 Updating <notifications> ...
2018-11-09T11:02:17+00:00 Updated <notifications> to 0.4.0
2018-11-09T11:02:17+00:00 Updating <provisioning_api> ...
2018-11-09T11:02:17+00:00 Updated <provisioning_api> to 0.5.0
2018-11-09T11:02:19+00:00 Repair step: Repair mime types
2018-11-09T11:02:19+00:00 Repair step: Detect file cache entries with path that does not match parent-child relationships
2018-11-09T11:02:19+00:00 Repair step: Generate ETags for file where no ETag is present.
2018-11-09T11:02:19+00:00 Repair info: ETags have been fixed for 0 files/folders.
2018-11-09T11:02:19+00:00 Repair step: Clean tags and favorites
2018-11-09T11:02:19+00:00 Repair info: 0 tags of deleted users have been removed.
2018-11-09T11:02:19+00:00 Repair info: 0 tags for delete files have been removed.
2018-11-09T11:02:19+00:00 Repair info: 0 tag entries for deleted tags have been removed.
2018-11-09T11:02:19+00:00 Repair info: 0 tags with no entries have been removed.
2018-11-09T11:02:19+00:00 Repair step: Drop old database tables
2018-11-09T11:02:19+00:00 Drop old database tables
2018-11-09T11:02:19+00:00 
                                                    2018-11-09T11:02:19+00:00  Done
 28/28 [============================] 100%2018-11-09T11:02:19+00:00 
2018-11-09T11:02:19+00:00 Repair step: Drop old background jobs
2018-11-09T11:02:19+00:00 Repair step: Remove getetag entries in properties table
2018-11-09T11:02:19+00:00 Repair info: Removed 0 unneeded "{DAV:}getetag" entries from properties table.
2018-11-09T11:02:19+00:00 Repair step: Repair outdated OCS IDs
2018-11-09T11:02:19+00:00 Repair step: Repair invalid shares
2018-11-09T11:02:19+00:00 Repair step: Remove old share propagation app entries
2018-11-09T11:02:19+00:00 Repair step: Move user avatars outside the homes to the new location
2018-11-09T11:02:19+00:00 Move user avatars outside the homes to the new location
                                                    2018-11-09T11:02:19+00:00  Done
 1/1 [============================] 100%2018-11-09T11:02:19+00:00 
2018-11-09T11:02:19+00:00 Repair step: Remove shares of a users root folder
2018-11-09T11:02:19+00:00 Repair step: Repair unmerged shares
2018-11-09T11:02:19+00:00 Repair step: Disable extra themes
2018-11-09T11:02:19+00:00 Repair step: Repair sub shares
2018-11-09T11:02:19+00:00 Starting code integrity check...
2018-11-09T11:02:40+00:00 Finished code integrity check
2018-11-09T11:02:40+00:00 Update successful
2018-11-09T11:02:40+00:00 Maintenance mode is kept active
2018-11-09T11:02:40+00:00 Reset log level
➜  owncloud4
```
- Now try to run the test for `SystemTagManagerTest.php`
```
➜  tests ../lib/composer/phpunit/phpunit/phpunit --configuration phpunit-autotest.xml lib/SystemTag/SystemTagManagerTest.php 
Cannot load Xdebug - it was already loaded
PHPUnit 5.7.27 by Sebastian Bergmann and contributors.

Runtime:       PHP 7.2.11-4+ubuntu18.04.1+deb.sury.org+1 with Xdebug 2.6.1
Configuration: /home/sujith/test/owncloud4/tests/phpunit-autotest.xml

.............................................................     61 / 61 (100%)

Time: 34.15 seconds, Memory: 14.00MB

OK (61 tests, 170 assertions)
➜  tests
```
- Failure is not observed.

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Technical debt
- [ ] Tests

## Checklist:

- [ ] Code changes
- [ ] Unit tests added
- [ ] Acceptance tests added
- [ ] Documentation ticket raised: <link> 

## Open tasks:

- [ ] Backport (if applicable set "backport-request" label and remove when the backport was done)

### CODE REVIEW

**Review by DeepDiver1975 — CHANGES_REQUESTED** [review]

**Comment by sharidas** [comment]
> > Update the unique index for the systemtag table.
> 
> This quite obvious ... but how is this fixing the issue? What is the issue? Please explain. THX

### Issue ###
- Let us consider the dataprovider https://github.com/owncloud/core/blob/master/tests/lib/SystemTag/SystemTagManagerTest.php#L95-L102 for the test https://github.com/owncloud/core/blob/master/tests/lib/SystemTag/SystemTagManagerTest.php#L110
- The moment https://github.com/owncloud/core/blob/master/tests/lib/SystemTag/SystemTagManagerTest.php#L101 dataset is taken for execution, the database snapshot looks like, shown below:
```
sqlite> select * from oc_systemtag;
id          name        visibility  editable    assignable
----------  ----------  ----------  ----------  ----------
315         one         0           0           1         
316         one         1           0           1         
317         one         0           1           1         
318         one         1           1           1         
319         two         0           0           0         
sqlite>
```
- When the code tries to create a new tag for https://github.com/owncloud/core/blob/master/tests/lib/SystemTag/SystemTagManagerTest.php#L101, the new entry we expect will be like:
```
320         two         0           0           1
```
This conflicts with the unique index. For a moment just look at the column `name`, `visibility` and `editable` column only, then rows with id 319 and planned insertion (id 320) are duplicates!!! And causes the failure while executing the last dataprovider in the test. 
- The schema of the table is as shown below:
```
sqlite> .schema oc_systemtag
CREATE TABLE oc_systemtag (id INTEGER NOT NULL PRIMARY KEY AUTOINCREMENT, name VARCHAR(64) DEFAULT '' NOT NULL, visibility SMALLINT DEFAULT 1 NOT NULL, editable SMALLINT DEFAULT 1 NOT NULL, assignable INTEGER DEFAULT 0 NOT NULL);
CREATE UNIQUE INDEX tag_ident ON oc_systemtag (name, visibility, editable);
sqlite>
```
the column `assignable` is missing in the unique index.

### How this fix solves the problem ###
- When `assignable` column is added along with `name`, `visibility` and `editable` in the unique index, the issue caused by adding:
```
320         two         0           0           1
```
gets rectified because row id `319` and `320` are different. Only difference is in the assignable column ( 0 and 1).


#### Reason for modifying the migration ####
The reason I added the change to the existing migration was, while migrating from any lower version to the this version, the migration will have change to add the `assignable` to the oc_systemtag and it will also have the change to add `assignable` to the unique index. But then the question is what if this migration is already applied by a user. In that case the pr will have no effect. Thanks for the pointer. I will update the PR by creating a new migration code.

**Comment by sharidas** [comment]
@DeepDiver1975 

I have updated the code by
- [x] creating a new migration for adding column `assignable` to unique index. This should also help users who have already applied the migration for the change made to systemtags.
- [x] Verified in my machine that this code applies cleanly for the migration from 8.2.11 to daily-stable10-qa,
- [x] Verified the failed test do pass locally on my machine.

Kindly review my change.

File: core/Migrations/Version20181113071753.php [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+/**
+ * Add assignable column to unique index in oc_systemtags table
+ */
+class Version20181113071753 implements ISchemaMigration {
+
+	public function changeSchema(Schema $schema, array $options) {
+		$this->prefix = $options['tablePrefix'];
+		$table = $schema->getTable("{$this->prefix}systemtag");
+		if ($schema->hasTable("{$this->prefix}systemtag")) {
+			if ($table->hasColumn('assignable')) {
```
Comment (DeepDiver1975) [inline core/Migrations/Version20181113071753.php]: why would the table not have this column?

Comment (DeepDiver1975) [reply core/Migrations/Version20181113071753.php]: The column is added in a previous migration ...

**Review by DeepDiver1975 — CHANGES_REQUESTED** [review]

File: core/Migrations/Version20181113071753.php (hunk 1 above)
Comment (sharidas) [reply core/Migrations/Version20181113071753.php]: I totally agree with you. I was trying to be little defensive with the code. I will update the code by removing the check.

**Comment by sharidas** [comment]
- [x] Updated the change by removing the additional check made in the previous version of the code for the column. Addressed the change requested.

Kindly review my change

**Review by PVince81 — APPROVED** [review]
:+1:

**Comment by PVince81** [comment]
@sharidas please backport

**Comment by sharidas** [comment]
Backport to stable10 PR https://github.com/owncloud/core/pull/33522

**Pull request merged**